_posts/2019-05-20-server-rendering-responsively.md

```diff
@@ -0,0 +1,203 @@
+
+We really wanted to not maintain our own list of user agents. Alas, we found that none of the existing user agent
+detection libraries surfaced all the information we needed in a single resource. We needed to know the minimum
+width for a browser on a given device, and if it was resizable, and to what dimensions it was resizable. If any
+existing libraries _did_ have this data, they didn't provide it to us easily.
+
+So we did some experimentation, given the browsers and devices we knew we needed to support. And yeah...we
+(reluctantly) created our own user-agent detection library,
+[@artsy/detect-responsive-traits](https://github.com/artsy/detect-responsive-traits). We're using this to determine
+if your browser is likely going to use only the mobile breakpoint of our app, in which case we don't have to also
+render the desktop version. The library is currently targeting only the browsers and devices we support on
+[artsy.net](artsy.net), but
+[we're always open to contributions](https://github.com/artsy/detect-responsive-traits)!
+
+We aren't doing any detection of desktop browsers. They are more resizable than mobile browsers, and we are more
+concerned with mobile users getting less content sent over their 3G connection.
+
+## Why didn't you \_\_\_?
+
+Those are our primary tools for combining SSR with responsive design! They work well for us. We considered many
+many other options along the way. Here are a couple:
+
+### Use [`react-media`](https://github.com/ReactTraining/react-media) or [`react-responsive`](https://github.com/contra/react-responsive)
+
+We investigated both `react-media` and `react-responsive`, but found that they didn't approach the SSR side of the
+problem as deeply as we needed.
+
+We also weren't fans of the imperative API in `react-media`. We started with a similar API when building
+`react-responsive-media`, but learned that it limited our ability to render all breakpoints from the server. The
+API requires only one branch to be rendered, and this caused the aforementioned mismatches when the app would
+hydrate.
+
+With `react-responsive`, we didn't like that it relied on user agent detection as its primary method of handling
+SSR.
+
+### Rely solely on CSS
+
+As mentioned before, we render all breakpoints from the server and hide the non-matching branches with CSS. The
+issue with this approach, when combined with React, is that after hydration you have many components that are
+mounted and rendered unnecessarily. There's a performance hit you take for rendering components your user isn't
+seeing, but even worse is the potential for duplicate side-effects.
+
+Imagine a component that, when rendered, emits a call to an analytics service. If this component exists in both a
+mobile and desktop branch, you're now double-stuffing your analytics. Hopefully your analytics service is smart
+enough to count both calls, but it's still a bad idea to duplicate components that have side-effects.
```

Comment:
I feel like here we should repeat that we don’t render these branches at all on the client, to avoid this issue.